### PR TITLE
Add DHCP config

### PR DESCRIPTION
**Pull request by pinpox**

### CODE REVIEW

File: dhcp.nix [hunk 1]
```diff
@@ -3,11 +3,81 @@ let
   cfg = config.pi-router;
 in
 {
+  systemd.network.enable = true;
 
-  # TODO Setup DHCP. Maybe KEA?
+  networking = {
+    useDHCP = false;
+    nameservers = cfg.dnsServers;
+    firewall.interfaces."${cfg.interfaces.lan.name}".allowedUDPPorts = [ 67 68 ];
```
Comment (MayNiklas) [inline dhcp.nix]: I'm pretty sure we only have to open UDP port 67.
Can we try that?

File: dhcp.nix [hunk 2]
```diff
@@ -3,11 +3,81 @@ let
...
+      # Address=10.1.1.1/24
+    };
+
+    # [DHCPServer] section of the unit.
+    dhcpServerConfig = {
+      # EmitDNS = false;
+      # PoolOffset = 50;
+      # EmitDNS = "yes";
+      # ServerAddress = "192.168.101.1/24";
+      DNS = cfg.dnsServers;
```
Comment (MayNiklas) [inline dhcp.nix]: The DHCP server is not advertising our Unbound server as a DNS Server.
We should use `cfg.interfaces.lan.ip` here.

File: nftables.nix [hunk 3]
```diff
@@ -23,87 +23,51 @@ let cfg = config.pi-router.interfaces; in
...
+      # routers may also want: mld-listener-query, nd-router-solicit
+      ip6 nexthdr icmpv6 icmpv6 type {
+        destination-unreachable,
+        packet-too-big,
+        time-exceeded,
+        parameter-problem,
+        nd-router-advert,
+        nd-neighbor-solicit,
+        nd-neighbor-advert
+      } accept
```
Comment (MayNiklas) [inline nftables.nix]: We should remove icmpv6 rules until we need them.

File: nftables.nix [hunk 4]
```diff
@@ -23,87 +23,51 @@ let cfg = config.pi-router.interfaces; in
...
-          chain POSTROUTING {
-              type nat hook postrouting priority srcnat; policy accept;
+    firewall.filterForward = true;
 
-              # NAT between LAN and WAN
-              oifname $WAN_IFC ip saddr $LAN_NET counter masquerade
-          }
-      }
+    firewall.extraForwardRules = ''
+      # type filter hook forward priority 0;
```
Comment (MayNiklas) [inline nftables.nix:74]: Can we include that without it failing?

File: nftables.nix [hunk 5]
```diff
@@ -1,4 +1,4 @@
-{ config, pkgs, lib, ... }:
+{ config, ... }:
 let cfg = config.pi-router.interfaces; in
 {
```
Comment (MayNiklas) [inline nftables.nix:12]: Can we maybe add a comment here, including the command to build the nftables config?

File: configuration.nix [hunk 6]
```diff
@@ -12,40 +12,45 @@ in
...
-            type = types.str;
-            description = "Interface used for LAN";
-            default = "end0";
-          };
+    dnsServers = mkOption {
+      type = types.listOf types.str;
+      description = "List of DNS servers";
+      default = [ "1.1.1.1" "1.0.0.1" ];
+      example = [ "9.9.9.9" "1.0.0.1" ];
+    };
```
Comment (MayNiklas) [inline configuration.nix]: We should discuss this option in combination with the Unbound server.
Our router provides a DNS server our clients should use.
I still live the idea of a DNS option, but we should implement a different behavior.

**Review by MayNiklas — COMMENTED** [review]
We are getting very close!

File: dhcp.nix (hunk 2 above)
Comment (MayNiklas) [reply dhcp.nix]: This works:

```suggestion
      DNS = [ cfg.interfaces.lan.ip ];
```

File: dhcp.nix (hunk 1 above)
Comment (MayNiklas) [reply dhcp.nix]: Can confirm - still works:

```suggestion
    firewall.interfaces."${cfg.interfaces.lan.name}".allowedUDPPorts = [ 67 ];
```

File: nftables.nix (hunk 4 above)
Comment (MayNiklas) [reply nftables.nix:74]: For some reason we can't. Wasn't a problem with my old config... 🤔

File: nftables.nix (hunk 3 above)
Comment (MayNiklas) [reply nftables.nix:48]:
```suggestion
      # ICMP
      ip protocol icmp icmp type {
        destination-unreachable,
        router-advertisement,
        time-exceeded,
        parameter-problem
      } accept

      # count and drop any other traffic
      counter drop
    '';
```

Comment (MayNiklas) [reply nftables.nix:48]:
```suggestion
      # ICMP
      ip protocol icmp icmp type {
        destination-unreachable,
        router-advertisement,
        time-exceeded,
        parameter-problem
      } accept

      # count and drop any other traffic
      counter drop
    '';
```

File: dhcp.nix [hunk 7]
```diff
@@ -3,11 +3,81 @@ let
   cfg = config.pi-router;
 in
 {
+  systemd.network.enable = true;
 
-  # TODO Setup DHCP. Maybe KEA?
+  networking = {
+    useDHCP = false;
+    nameservers = cfg.dnsServers;
```
Comment (MayNiklas) [inline dhcp.nix]: This does nothing.
Our local Unbound server is being used (see `nslookup google.de`)
-> this only would apply when Unbound is disabled.
Maybe we should make this more clear?

File: nftables.nix (hunk 3 above)
Comment (pinpox) [reply nftables.nix]: commented out for now.

File: dhcp.nix (hunk 7 above)
Comment (pinpox) [reply dhcp.nix]: Removed.

File: configuration.nix (hunk 6 above)
Comment (pinpox) [reply configuration.nix]: removed for now.

File: nftables.nix (hunk 4 above)
Comment (pinpox) [reply nftables.nix:74]: No idea why. Let's leave it commented out for now.

File: nftables.nix (hunk 5 above)
Comment (pinpox) [reply nftables.nix:12]: Added command. Not pretty but it works.

**Pull request merged**